### ISSUE TITLE
Dispatch events to parent when can't swipe to a direction

### ISSUE THREAD

**Issue by LouisCAD** [h6ah4i/android-advancedrecyclerview#69]
Hi!
I am using your app with a ViewPager, and I would like to swipe the viewpager to the left, and swipe to dismiss list items to the right, but currently, both touch events are consumed. Although that the item isn't swiped to the left (because I set it to not), the ViewPager doesn't respond to swipe events. Could you fix this please?
Thanks.
All the best

**Comment by LouisCo** [comment]
Hi again!

I modified myself the version 0.7.2 of the library to fix this because my app needed fix for this as I'm using Google Now Launcher like card dismiss with two swipe-able tabs. I don't have skills enough to submit a proper pull request for now, so I'm posting here the code I modified to make this enhancement working.

Here's the code:
In **``swipeable/RecyclerViewSwipeManager.java``** I added the following method:

```java
private boolean isSwipeAllowed(RecyclerView.ViewHolder holder, int itemPosition, MotionEvent e) {
        int swipeDirection = mAdapter.getSwipeReactionType(holder, itemPosition, Math.round(e.getX()), Math.round(e.getY()));
        if (swipeDirection == REACTION_CAN_SWIPE_BOTH) return true;
        float xMoved = e.getX() - e.getHistoricalX(0);
        if (xMoved > 0f && SwipeReactionUtils.canSwipeRight(swipeDirection)) return true;
        else if (xMoved < 0f && SwipeReactionUtils.canSwipeLeft(swipeDirection)) return true;
        else if (xMoved == 0f) {
            Log.wtf(TAG, "moved with no move??");
            return true;
        } else return false;
    }
```

I call this method in the same class, in method ``private boolean handleActionMoveWhileNotSwiping(RecyclerView rv, MotionEvent e)``.
Just above this few lines of code at the end of the method:
```java
if (LOCAL_LOGD) {
    Log.d(TAG, "swiping started");
}

startSwiping(rv, e, holder, itemPosition);
```
I added
```java
if (!isSwipeAllowed(holder, itemPosition, e)) return false;
```

I also realized that Switch swiping wasn't working properly because ``RecyclerViewSwipeManager`` intercepted all touch events, even if a swipe-able widget is touched. So I added the code to prevent touch interception when ``onRequestDisallowInterceptTouchEvent(boolean disallowIntercept)`` is called.

To do so, still in the same class, I added a member variable:
```java
private boolean mDisallowIntercept = false;
```
and I used it in the callback methods defined in the constructor:
```java
public RecyclerViewSwipeManager() {
        mInternalUseOnItemTouchListener = new RecyclerView.OnItemTouchListener() {
            @Override
            public boolean onInterceptTouchEvent(RecyclerView rv, MotionEvent e) {
                if (mDisallowIntercept) {
                    Log.wtf(TAG, new RuntimeException("onInterceptTouchEvent not allowed"));
                    return false;
                }
                return RecyclerViewSwipeManager.this.onInterceptTouchEvent(rv, e);
            }

            @Override
            public void onTouchEvent(RecyclerView rv, MotionEvent e) {
                if (mDisallowIntercept) {
                    Log.wtf(TAG, new RuntimeException("onTouchEvent not allowed"));
                }
                RecyclerViewSwipeManager.this.onTouchEvent(rv, e);
            }

            @Override
            public void onRequestDisallowInterceptTouchEvent(boolean disallowIntercept) {
                mDisallowIntercept = disallowIntercept;
            }
        };
        mItemSlideAnimator = new ItemSlidingAnimator();
        mVelocityTracker = VelocityTracker.obtain();
    }
```
It then allows me to disable item dismiss when I get touch action down on my swipe-able View (Switch in this case) and enable it again on touch action up.

I hope my work will help you, and will help developers using your library :wink:.

Please, let me know when you release this enhancement so I use your version again.

All the best!

**Comment by LouisCo** [comment]
Can't you include my code in the next release?

**Comment by h6ah4i** [comment]
Hi. Sorry for the late reply :disappointed: I'm going to include this enhancement in the next release.

**Comment by h6ah4i** [comment]
Hi. I have just published v0.8.0 and it includes a new demo ["Swipeable with ViewPager"](https://github.com/h6ah4i/android-advancedrecyclerview/tree/023f6c0cbcecc9f48f38622a7e8937b10b3d858c/example/src/main/java/com/h6ah4i/android/example/advrecyclerview/demo_s_viewpager), but it is implemented by a different approach. 


First, I have introduced some new flags which can control how swipe starts.

- `REACTION_MASK_START_SWIPE_LEFT`
- `REACTION_MASK_START_SWIPE_RIGHT`
- `REACTION_MASK_START_SWIPE_UP`
- `REACTION_MASK_START_SWIPE_DOWN`

:point_right: If these flags are specified, swipe does not start if finger moves to the direction.

- `REACTION_START_SWIPE_ON_LONG_PRESS`

:point_right: If this flag is specified, swipe starts on long press.

If you want to permit swipe LEFT only and do not want to consume touch event other directions, specify as follows;

```java
REACTION_CAN_SWIPE_LEFT |
REACTION_MASK_START_SWIPE_RIGHT | 
REACTION_MASK_START_SWIPE_UP | 
REACTION_MASK_START_SWIPE_DOWN
```

---

However, it requires **one more trick** to work with ViewPager properly.
```xml
<ViewPager>
    <RecyclerView>
        <!-- list items -->
    </RecyclerView>
</ViewPager>

```

When a RecyclerView is placed in a ViewPager, the order of `onInterceptTouchEvent()` method call is;

1. `ViewPager`
2. `RecyclerView`

Then, we need to avoid the ViewPager consumes touch event if user swipes to LEFT direction. I made a custom view [`PagerSwipeItemFrameLayout`](https://github.com/h6ah4i/android-advancedrecyclerview/blob/023f6c0cbcecc9f48f38622a7e8937b10b3d858c/example/src/main/java/com/h6ah4i/android/example/advrecyclerview/demo_s_viewpager/PagerSwipeItemFrameLayout.java) which can control return value of the [`canScrollHorizontally()`](https://github.com/h6ah4i/android-advancedrecyclerview/blob/023f6c0cbcecc9f48f38622a7e8937b10b3d858c/example/src/main/java/com/h6ah4i/android/example/advrecyclerview/demo_s_viewpager/PagerSwipeItemFrameLayout.java#L40) method.

```xml
<ViewPager>
    <RecyclerView>
        <!-- list items -->
        <PagerSwipeItemFrameLayout />
        <PagerSwipeItemFrameLayout />
        <PagerSwipeItemFrameLayout />
         ...
    </RecyclerView>
</ViewPager>
```

```java
class PagerSwipeItemFrameLayout extends FrameLayout {
    ...
    @Override
    public boolean canScrollHorizontally(int direction) {
        if (mCanSwipeLeft && direction > 0) {
            // return true to avoid view pager consume swipe left (= scroll right) touch events
            return true;
        }
        if (mCanSwipeRight && direction < 0) {
            // return true to avoid view pager consume swipe right (= scroll left) touch events
            return true;
        }

        return false;
    }
   ...
}
```

---

Please check [the implementation](https://github.com/h6ah4i/android-advancedrecyclerview/tree/023f6c0cbcecc9f48f38622a7e8937b10b3d858c/example/src/main/java/com/h6ah4i/android/example/advrecyclerview/demo_s_viewpager)) fore more details.

Thanks.

**Comment by LouisCAD** [comment]
Thank you very much! I'm going to leave my workaround version I showed above to get this new, awesome version! You rocks! The new GridLayout version is also very good! Just some more animations to make drag n drop smoother, and it will be perfect (or very close :) ). It could be used in a Launcher for example (I think RecyclerView could improve performance on low-end devices and prevent some launcher redraws)
